src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +119,213 @@ def snapshot(self) -> Iterable[JobInfo]:
+            hash_values.extend(self._retrieve_hash_values_from_condition_task(task.condition_task))
+        if task.run_job_task is not None and task.run_job_task.job_id is not None:
+            hash_values.append(str(task.run_job_task.job_id))
+        if task.git_source is not None:
+            hv = self._retrieve_hash_values_from_git_source(task.git_source)
+            hash_values.extend(hv)
+
+        return [str(value) for value in hash_values if value is not None]
+
+    def check_spark_conf(self, conf: dict[str, str], source: str) -> list[str]:
+        failures: list[str] = []
+        for key in conf.keys():
+            if any(re.search(regexp, key) for regexp in self._FS_LEVEL_KEY_REGEXP):
+                failures.append(f"unsupported config property: {key}")
+
+        failures.extend(super().check_spark_conf(conf, source))
+        return failures
+
+    def check_cluster_failures(self, cluster: ClusterDetails, source: str) -> list[str]:
+        failures: list[str] = []
+        if cluster.aws_attributes and cluster.aws_attributes.instance_profile_arn:
+            failures.append(f"using instance profile: {cluster.aws_attributes.instance_profile_arn}")
+
+        failures.extend(super().check_cluster_failures(cluster, source))
+        return failures
+
+    @staticmethod
+    def _task_needs_compatibility_check(task: RunTask) -> bool:
+        """
+        # we recognize a task as a potentially incompatible one if:
+        # 1. cluster is not configured with data security mode
+        # 2. cluster's DBR version is greater than 11.3
+        """
+        if task.new_cluster is not None and task.new_cluster.data_security_mode is None:
+            compatability = spark_version_compatibility(task.new_cluster.spark_version)
+            return compatability == "supported"
+
+        return False
+
+    def _assess_job_runs(self, submit_runs: Iterable[BaseRun]) -> Iterable[SubmitRunInfo]:
+        """
+        Assessment logic:
+        1. For each submit run, we analyze all tasks inside this run.
+        2. Per each task, we calculate a unique hash based on the _retrieve_hash_values_from_task function
+        3. Then we coalesce all task hashes into a single hash for the submit run
+        4. Coalesce all runs under the same hash into a single pseudo-job
+        5. Return a list of pseudo-jobs with their assessment results
+        """
+        result: dict[str, SubmitRunInfo] = {}
+        runs_per_hash: dict[str, list[int | None]] = {}
+
+        for submit_run in submit_runs:
+            hashable_items = []
+            all_tasks: list[RunTask] = submit_run.tasks if submit_run.tasks is not None else []
+            failures_per_task: dict[str, list[str]] = {}
+
+            for task in sorted(all_tasks, key=lambda x: x.task_key if x.task_key is not None else ""):
+                hashable_items.extend(self._retrieve_hash_values_from_task(task))
+
+                if self._task_needs_compatibility_check(task):
```

Comment:
Rewrite with "if not ...: continue", the code is too deeply nested and unreadable